### PR TITLE
Orbidder: bidfloor currency handling

### PR DESCRIPTION
**Pull request by andreasmaurer0210**
The Orbidder Backend only supports EUR. This PR deals with Bid Floor Currency conversion into EUR as requested by mail from the Prebid Server committee.

### CODE REVIEW

File: adapters/orbidder/orbidder.go [hunk 1]
```diff
@@ -80,8 +88,25 @@ func preprocess(imp *openrtb2.Imp) error {
 	return nil
 }
 
+func preprocessBidFloorCurrency(imp *openrtb2.Imp, reqInfo *adapters.ExtraRequestInfo) error {
+	// we expect every currency related data to be EUR
+	if imp.BidFloorCur == "" {
+		imp.BidFloorCur = "EUR"
```
Comment (guscarreon) [inline adapters/orbidder/orbidder.go:94]: Given that `imp.BidFloorCur` ends up being set to `"EUR"` in line 104 anyways, we could remove this `if` statement altogether.

**Review by guscarreon — COMMENTED** [review]
Your code looks pretty good. I left a nitpick below but not a deal breaker.

**Review by guscarreon — APPROVED** [review]

File: adapters/orbidder/orbidder.go [hunk 2]
```diff
@@ -55,14 +36,41 @@ func (rcv *OrbidderAdapter) MakeRequests(request *openrtb2.BidRequest, reqInfo *
...
 
-func preprocess(imp *openrtb2.Imp) error {
+// getValidImpressions validate imps and check for bid floor currency. Convert to EUR if necessary
+func getValidImpressions(request *openrtb2.BidRequest, reqInfo *adapters.ExtraRequestInfo) ([]openrtb2.Imp, []error) {
+	var errs []error
+	var validImps []openrtb2.Imp
+
+	// check if imps exists, if not return error and do send request to orbidder.
+	if len(request.Imp) == 0 {
+		return nil, []error{&errortypes.BadInput{
```
[the remark this answers is not in the document]
Comment (andreasmaurer0210) [reply adapters/orbidder/orbidder.go]: Thanks @guscarreon for your quick review and useful advice. We removed the check for an empty Imp slice already.

**Review by guscarreon — DISMISSED** [review]
Looks good to me. I believe getting rid of the conditional in line 86 makes for an easier read but I'm good either way and I'm approving anyways
```
84   func preprocessBidFloorCurrency(imp *openrtb2.Imp, reqInfo *adapters.ExtraRequestInfo) error {
85       // we expect every currency related data to be EUR
86 -     if imp.BidFloorCur == "" {
87 -         imp.BidFloorCur = "EUR"
88 -     }
89 -
90 -     if imp.BidFloor > 0 && strings.ToUpper(imp.BidFloorCur) != "EUR" {
   +     if imp.BidFloor > 0 && (imp.BidFloorCur == "" || strings.ToUpper(imp.BidFloorCur) != "EUR") {
91           if convertedValue, err := reqInfo.ConvertCurrency(imp.BidFloor, imp.BidFloorCur, "EUR"); err != nil {
92               return err
93           } else {
94               imp.BidFloor = convertedValue
95           }
96       }
97       imp.BidFloorCur = "EUR"
98       return nil
99   }
adapters/orbidder/orbidder.go
```

Thanks @andreasmaurer0210 for addressing my feedback.

File: adapters/orbidder/orbidder.go [hunk 3]
```diff
@@ -55,14 +36,34 @@ func (rcv *OrbidderAdapter) MakeRequests(request *openrtb2.BidRequest, reqInfo *
...
+			continue
+		}
+		validImps = append(validImps, imp)
+	}
+	return validImps, errs
+}
+
+func preprocessExtensions(imp *openrtb2.Imp) error {
 	var bidderExt adapters.ExtImpBidder
 	if err := json.Unmarshal(imp.Ext, &bidderExt); err != nil {
```
Comment (hhhjort) [inline adapters/orbidder/orbidder.go:68]: Are you just testing that unmarshalling `imp.ext` does not produce an error? Seems that you are throwing away the result.

**Comment by mansinahar** [comment]
@andreasmaurer0210 Friendly ping to see if you have a chance to address the comment above

**Comment by andreasmaurer0210** [comment]
> @andreasmaurer0210 Friendly ping to see if you have a chance to address the comment above

@mansinahar Sry, for my late reply. I will take care of the comments.

File: adapters/orbidder/orbidder.go (hunk 3 above)
Comment (andreasmaurer0210) [reply adapters/orbidder/orbidder.go:68]: Yes, you are right. We only checking for errors in unmarshalling `imp.ext`. In the passt we had some problems with invalid Extensions send to our adapter, so we find this useful. We added soem testecases for the preprocessing.

File: adapters/orbidder/orbidder_test.go [hunk 4]
```diff
@@ -32,3 +50,116 @@ func TestJsonSamples(t *testing.T) {
...
+	for name, tc := range testCasesCurrency {
+		t.Run(name, func(t *testing.T) {
+			imp := tc.imp
+			mockConversions := &mockCurrencyConversion{}
+			tc.setMock(&mockConversions.Mock)
+			extraRequestInfo := adapters.ExtraRequestInfo{
+				CurrencyConversions: mockConversions,
+			}
+
+			err := preprocessBidFloorCurrency(&imp, &extraRequestInfo)
```
Comment (mansinahar) [inline adapters/orbidder/orbidder_test.go:146]: You need to call `mockCurrencyConversion.AssertExpectations(t)` to make sure that the calls that you're expecting to the currency converter implementation are called with the args you have specified in `setMock` in those test cases.

File: adapters/orbidder/orbidder_test.go [hunk 5]
```diff
@@ -32,3 +50,116 @@ func TestJsonSamples(t *testing.T) {
...
+			BidFloor:    1,
+			BidFloorCur: "EUR",
+		},
+		setMock: func(m *mock.Mock) {},
+		expectedImp: openrtb2.Imp{
+			BidFloor:    1,
+			BidFloorCur: "EUR",
+		},
+	},
+	"USD: bidfloor with currency": {
```
Comment (mansinahar) [inline adapters/orbidder/orbidder_test.go:92]: Can you please add another similar test case where the mock converter returns an error to make sure the error from the converter is properly propagated?

File: adapters/orbidder/orbidder_test.go (hunk 4 above)
Comment (andreasmaurer0210) [reply adapters/orbidder/orbidder_test.go:161]: Thank you for pointing this out to us. We added a call of `AsserrExpectations` in our test method. Always happy to learn new stuff :)

File: adapters/orbidder/orbidder_test.go (hunk 5 above)
Comment (andreasmaurer0210) [reply adapters/orbidder/orbidder_test.go:92]: Done.

**Review by mansinahar — APPROVED** [review]

**Comment by mansinahar** [comment]
Thanks @andreasmaurer0210 for addressing the comments. LGTM!

**Review by AlexBVolcy — APPROVED** [review]

**Pull request merged**